### ISSUE TITLE
Potential Issue: internal message handlers run in uncontrolled goroutines

### ISSUE THREAD

**Issue by OrMemphis** [memphisdev/memphis#321]
### Current behavior

the broker is subscribing to "memphis" sdk requests, when a message arrives a handler is invoked in a new goroutine.
Thus, a decent amount of these requests can potentially deteriorate the broker's performance

### Suggested solution

Consider using a goroutine pool with a predetermined limit for the amount of goroutines

maybe use [tunny](https://github.com/Jeffail/tunny)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [x] I agree to follow this project's Contribution Docs

**Comment by bedLad** [comment]
Is anyone working on this issue?

**Comment by yanivbh1** [comment]
@idanasulinmemphis @OrMemphis are you on it?

**Comment by OrMemphis** [comment]
I'm not planning to work on this any time soon

**Comment by idanasulinmemphis** [comment]
@bedLad hi glad you to see you want to jump in to make your first contribution for memphis. 
Please feel free to take this one, nobody else is dealing with it at the moment

**Comment by bedLad** [comment]
@idanasulinmemphis ok i will work on it

**Comment by idanasulinmemphis** [comment]
@bedLad great, have you already joined our Discord server?

https://memphis.dev/discord

**Comment by bedLad** [comment]
@OrMemphis can you please tell me where does the broker receive these sdk request

**Comment by idanasulinmemphis** [comment]
@bedLad https://github.com/memphisdev/memphis-broker/blob/master/server/memphis_sdk_handlers.go

**Comment by bedLad** [comment]
> @bedLad https://github.com/memphisdev/memphis-broker/blob/master/server/memphis_sdk_handlers.go

Ok so the above file has a func `initializeSDKHandlers()` which has go routine and other similar go routine
```
func createStationHandler(s *Server) simplifiedMsgHandler {
	return func(c *client, subject, reply string, msg []byte) {
		go s.createStationDirect(c, reply, copyBytes(msg))
	}
}
```
the keyword 'go' which enables the goroutine to run concurrently while new functions are being called. 

*******************
So the issue is that 
too many of these function calls can deteriorate the performance of the broker.

**So I should limit these concurrent above mentioned func & similar func executions with the help of tunny.**

Please correct me if I misinterpreted the problem statement. I'm new to all this.

**Comment by OrMemphis** [comment]
@bedLad you got it

**Comment by bedLad** [comment]
Hello,
I used [https://github.com/zenthangplus/goccm] to limit the goroutines
Could you tell me how i should go about testing this

**Comment by OrMemphis** [comment]
Hi, first, thanks for the effort 😄 
Basically, for every message that is a "memphis" request we'll run a new goroutine (to handle the request), so you can try watching the memory footprint of the broker with or w/o your additions.
To do that: 
1. run the broker
2. with a memphis SDK (i.e. memphis.go or memphis.py...) create a lot of Stations/Consumers/Producers. each such "creation" will be translated to a memphis request that we handle in a new goroutine.
You should be able to notice the memory the process consumes grows with the amount of requests.
Hopefully with you fix it should be constant regardless of request amount.

**Comment by bedLad** [comment]
Sorry I was a little busy with uni exams. Will start working on this now.

**Comment by bedLad** [comment]
Hello,
I intitated the memphis broker and used memphis.go SDK for testing.

memphis.go SDK is added below.

I made 20 stations, producers & consumers. But I have no idea where to look for the memory that is being used by all the goroutines.
I will also add my station screen at the bottom.

Please help me.
```
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/memphisdev/memphis.go"
)

func main() {
	for i := 0; i < 20; i++ {
		ch := strconv.Itoa(i)
		go station(ch)
	}
	var input string
	fmt.Scanln(&input)
}

func station(ch string) {
	conn, err := memphis.Connect("localhost", "chadvik", "memphis")
	if err != nil {
		os.Exit(1)
	}
	defer conn.Close()
	p, err := conn.CreateProducer("station"+ch, "rutvik"+ch)

	hdrs := memphis.Headers{}
	hdrs.New()
	err = hdrs.Add("This message says : 'Hello Memphis'", "The message was from : '"+"rutvik"+ch+"'")

	if err != nil {
		fmt.Errorf("Header failed: %v", err)
		os.Exit(1)
	}

	err = p.Produce([]byte("You have a message!"), memphis.MsgHeaders(hdrs))

	if err != nil {
		fmt.Errorf("Produce failed: %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer("station"+ch, "vatsa"+ch, memphis.PullInterval(15*time.Second))

	if err != nil {
		fmt.Printf("Consumer creation failed: %v", err)
		os.Exit(1)
	}

	handler := func(msgs []*memphis.Msg, err error) {
		if err != nil {
			fmt.Printf("Fetch failed: %v", err)
			return
		}

		for _, msg := range msgs {
			fmt.Println(string(msg.Data()))
			msg.Ack()
			headers := msg.GetHeaders()
			fmt.Println(headers)
		}
	}

	consumer.Consume(handler)
	// The program will close the connection after 30 seconds,
	// the message handler may be called after the connection closed
	// so the handler may receive a timeout error
	time.Sleep(10 * time.Second)
}
```

![image](https://user-images.githubusercontent.com/47242147/212738472-3bb45dc7-e04c-4df0-926d-ec9fc8437192.png)

**Comment by idanasulinmemphis** [comment]
20 is not enough but try for example write an infinite for-loop creating stations.
Behind the scenes, every station creation request is getting handled in a separate go routine, so in that case, at some point, you will reach go routines overloading the system.
You can watch the memory/CPU with the top command for example in order to see if it happens.

Now a suggestion for fixing it can be creating a go routines pool in a fixed size, and finding a way to determine this size according to the machine Memphis run on

**Comment by bedLad** [comment]
So whenever I try to run beyond 110 goroutines simultaneously on my machine I get the following error.

```
map[$memphis_connectionId:24ec6b448d2d7c37bafe9019 $memphis_producedBy:rutvik85]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x20 pc=0xa7c435]

goroutine 20 [running]:
github.com/memphisdev/memphis%2ego.(*ProduceOpts).produce(0xc0006ff470, 0x0)
        C:/Users/vatsa.rutvik/go/pkg/mod/github.com/memphisdev/memphis.go@v0.1.7/producer.go:267 +0x55
github.com/memphisdev/memphis%2ego.(*Producer).Produce(0xc000338768?, {0xab1380?, 0xc000aca528}, {0xc000b0bf60, 0x1, 0xb4841a?})
        C:/Users/vatsa.rutvik/go/pkg/mod/github.com/memphisdev/memphis.go@v0.1.7/producer.go:241 +0x15c
main.station({0xb65a9b, 0x2})
        E:/go-proj/memphis-testing/main.go:38 +0x2ac
created by main.main
        E:/go-proj/memphis-testing/main.go:15 +0x25
exit status 2
```
Or sometimes its just the producer thats sending the data but the consumer stays dormant.
![image](https://user-images.githubusercontent.com/47242147/214120572-ff7124e8-5e6d-4914-b3a9-4d1b70dbf421.png)
![image](https://user-images.githubusercontent.com/47242147/214120791-83f288c2-709f-4b31-8552-d84940213b74.png)

**Comment by bedLad** [comment]
Should I limit the goroutines to about a 100 ? Or is there something I'm doing incorrectly ?

**Comment by idanasulinmemphis** [comment]
@bedLad ok, in order to test it out please don't send any messages, let's stick on the scenario of an infinite for loop which creates a station in each iteration.
The broker will handle each station creation in a separate Go routine.
The goal is that the broker will use tunny (goccm is not good enough for us after taking a look at it) in order to handle those requests in parallel instead of using the go keyword explicitly.

Please compare the CPU & memory usage before and after you are applying this fix and add it in here together with your PR.

Thanks again

**Comment by bedLad** [comment]
Hey could you elaborate on what I should do with this
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x20 pc=0x76c035]

goroutine 12 [running]:
github.com/memphisdev/memphis%2ego.(*ProduceOpts).produce(0xc022c37bc0, 0x0)
        C:/Users/vatsa.rutvik/go/pkg/mod/github.com/memphisdev/memphis.go@v0.1.7/producer.go:267 +0x55
github.com/memphisdev/memphis%2ego.(*Producer).Produce(0xc01e80c370?, {0x7a1380?, 0xc00e848000}, {0xc002405f60, 0x1, 0x0?})
        C:/Users/vatsa.rutvik/go/pkg/mod/github.com/memphisdev/memphis.go@v0.1.7/producer.go:241 +0x15c
main.station({0x848fe8, 0x1})
        E:/go-proj/memphis-testing/main.go:38 +0x274
created by main.main
        E:/go-proj/memphis-testing/main.go:15 +0x1d
exit status 2
```

Whenever I try to run the infinite for loop my program immediately crashes and I do not get the chance to compare the CPU & memory usage

**Comment by idanasulinmemphis** [comment]
Could you please share the code snippet causing this issue?

**Comment by bedLad** [comment]
the code runs ok for around 100 iterations but crashes with the above mentioned error message if I run an infinite for loop
```
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/memphisdev/memphis.go"
)

func main() {
	for i := 0; true; i++ {
		ch := strconv.Itoa(i)
		go station(ch)
	}
	var input string
	fmt.Scanln(&input)      // waits for the input if for loop is run for under 100 iterations
}

func station(ch string) {
	conn, err := memphis.Connect("localhost", "chadvik", "memphis")
	if err != nil {
		os.Exit(1)
	}
	defer conn.Close()
	p, err := conn.CreateProducer("station"+ch, "rutvik"+ch)

	hdrs := memphis.Headers{}
	hdrs.New()
	err = hdrs.Add("this is the key string", "this is the value string")

	if err != nil {
		fmt.Errorf("Header failed: %v", err)
		os.Exit(1)
	}

	err = p.Produce([]byte("Hello World!!!"), memphis.MsgHeaders(hdrs))

	if err != nil {
		fmt.Errorf("Produce failed: %v", err)
		os.Exit(1)
	}

	consumer, err := conn.CreateConsumer("station"+ch, "vatsa"+ch, memphis.PullInterval(15*time.Second))

	if err != nil {
		fmt.Printf("Consumer creation failed: %v", err)
		os.Exit(1)
	}

	handler := func(msgs []*memphis.Msg, err error) {
		if err != nil {
			fmt.Printf("Fetch failed: %v", err)
			return
		}

		for _, msg := range msgs {
			fmt.Println(string(msg.Data()))
			msg.Ack()
			headers := msg.GetHeaders()
			fmt.Println(headers)
		}
	}

	consumer.Consume(handler)
	// The program will close the connection after 15 seconds,
	// the message handler may be called after the connection closed
	// so the handler may receive a timeout error
	time.Sleep(15 * time.Second)
}


```

**Comment by idanasulinmemphis** [comment]
So first the same error doesn't reproduce on my machine, however, you are missing the point. The way to check it out is not to take the code example and make it run infinitely. You have to do something like this.
`connect()
for {
createStation()
}`

not in a separate go routine, as simple as an infinite for loop creating stations from the same connection. In the case at some point, you see Memphis is idle this is the problem we are talking about

Anyway, the reason I believe it is making issues on your local machine is that you are overloading your system with a ton of go routines

**Comment by bedLad** [comment]
Hey thank you for your suggestions. I will implement them and get update you as soon as my health gets better.

**Comment by idanasulinmemphis** [comment]
Fixed on version 0.4.5
